conans/test/unittests/tools/files_patch_test.py

```diff
@@ -0,0 +1,152 @@
+import os
+import unittest
+from textwrap import dedent
+
+from parameterized.parameterized import parameterized
+
+from conans.client.graph.python_requires import ConanPythonRequire
+from conans.client.loader import ConanFileLoader
+from conans.test.utils.test_files import temp_folder
+from conans.test.utils.tools import TestClient, TestBufferConanOutput,\
+    test_processed_profile
+from conans.util.files import save, load
+
+base_conanfile = '''
+from conans import ConanFile
+from conans.tools import patch, replace_in_file
+import os
+
+class ConanFileToolsTest(ConanFile):
+    name = "test"
+    version = "1.9.10"
+'''
+
+
+class ToolsFilesPatchTest(unittest.TestCase):
+
+    @parameterized.expand([(0, ), (1, )])
+    def test_patch_from_file(self, strip):
+        if strip:
+            file_content = base_conanfile + '''
+    def build(self):
+        patch(patch_file="file.patch", strip=%s)
+''' % strip
+            patch_content = '''--- %s/text.txt\t2016-01-25 17:57:11.452848309 +0100
++++ %s/text_new.txt\t2016-01-25 17:57:28.839869950 +0100
+@@ -1 +1 @@
+-ONE TWO THREE
++ONE TWO FOUR''' % ("old_path", "new_path")
+        else:
+            file_content = base_conanfile + '''
+    def build(self):
+        patch(patch_file="file.patch")
+'''
+            patch_content = '''--- text.txt\t2016-01-25 17:57:11.452848309 +0100
++++ text_new.txt\t2016-01-25 17:57:28.839869950 +0100
+@@ -1 +1 @@
+-ONE TWO THREE
++ONE TWO FOUR'''
+
+        tmp_dir, file_path, text_file = self._save_files(file_content)
+        patch_file = os.path.join(tmp_dir, "file.patch")
+        save(patch_file, patch_content)
+        self._build_and_check(tmp_dir, file_path, text_file, "ONE TWO FOUR")
+
+    def test_patch_from_str(self):
+        file_content = base_conanfile + '''
+    def build(self):
+        patch_content = \'''--- text.txt\t2016-01-25 17:57:11.452848309 +0100
++++ text_new.txt\t2016-01-25 17:57:28.839869950 +0100
+@@ -1 +1 @@
+-ONE TWO THREE
++ONE TWO DOH!\'''
+        patch(patch_string=patch_content)
+
+'''
+        tmp_dir, file_path, text_file = self._save_files(file_content)
+        self._build_and_check(tmp_dir, file_path, text_file, "ONE TWO DOH!")
+
+    def test_patch_new(self):
```

Comment:
This is the only test that is new, the other tests are just moved from the other location, sorry for mixing both in the same PR.

Reply:
I don't think that tests using the TestClient should be considered as unit testing. I would move this test to functional or do a real unit test of `tools.pathc()` in this case

Reply:
These tests were already in ``unittests``, I just gave them a dedicated file according to the layout.
They should be refactored and converted to real unit tests, it is not necessary that they use the TestClient.

In any case, I know it is not perfect, but I am not uncomfortable with some test in ``unittests`` that test something very, very specific like ``tools.patch()`` within a recipe. They shouldn't be the norm, but It is nice to have some complete test, and moving that single test to somewhere in ``functional`` would separate the logic too much.

One of the advantages of this structure of tests is knowing what to run when you are developing something. In this case, I am fixing a bug in ``tools.patch``, and it is very nice to be able to go to ``unittests/tools/files_patch_test`` and run all the relevant tests from there.

Reply:
Having everything in the same place is not a valid reason and knowing what to run is easy in this case: everything that has a ``tools.patch``, although I agree we should have "functional" or "integration" test using the TestClient too.

For me this invalidates the reason or organizing the tests and will eventually evolve to the same mess.